keps/draft-20190418-previous-values-in-update-plans-and-conditional-phases.md

```diff
@@ -0,0 +1,35 @@
+---
+kep-number: draft-20190418
+title: Previous values in update plans and conditional phases
+authors:
+  - "@djannot"
+owners:
+  - TBD
+editor: TBD
+creation-date: 2019-04-18
+status: provisional
+---
+
+# Previous values in update plans and conditional phases
+
+## Table of Contents
+
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+
+Created by [gh-md-toc](https://github.com/ekalinin/github-markdown-toc)
+
+## Summary
+
+In order to develop the logic behind an `update` `Plan`, you generally need to know what were the previous values.
+
+For example, if the number of replicas of a `StatefulSet` is updated, you often need to add a new `Plan` to reconfigure your cluster.
+
+But currently, you don't know what has changed. You don't know if the number of replicas has changed and if it has changed you don't know what was the previous value.
```

Comment:
For this example, cant the current StatefulSet be queried to identify what the current replicas is?

Reply:
I agree with @runyontr and I mentioned it in todays community meeting as well. We always have the previous values before a new change is being applied and then just need some sort of validator that follows a certain ruleset. Do we want that ruleset hard coded in to Kudo or provided by the developer shipped with the framework? Do we want a hybrid as in both? Or don't want it at all?

Thinking out loud here...

A way how we detect right now changes and cause actions is e.g. outlined here https://github.com/kudobuilder/kudo/blob/master/pkg/controller/planexecution/planexecution_controller.go#L490-L507 - in that case if we detect a change in replicas we adjust automatically the "previous" value with the "new" value, enabling scaling if you will. That is of course just a very detailed implementation and what I mentioned in the community meeting was more around the idea of having some sort of implementation that allows framework developers to customize more, e.g. like in Prometheus which is based on Prometheus expression language expressions to come up with their own custom rules.conf. Through the `UpdateFunc` ( https://github.com/kudobuilder/kudo/blob/master/pkg/controller/planexecution/planexecution_controller.go#L125-L128 ) we could always get triggered and then have developers provide their own set of rules, e.g. trigger plans or modify other parameters. Iterating on that, thinking sort of a rules engine living in `/pkg/util/rules` that evaluates/validates.